wrappers/services_test.go

```diff
@@ -163,6 +163,291 @@ WantedBy=multi-user.target
 	c.Check(s.sysdLog[1], DeepEquals, []string{"daemon-reload"})
 }
 
+func (s *servicesTestSuite) TestEnsureSnapServicesAdds(c *C) {
+	info := snaptest.MockSnap(c, packageHello, &snap.SideInfo{Revision: snap.R(12)})
+	svcFile := filepath.Join(s.tempdir, "/etc/systemd/system/snap.hello-snap.svc1.service")
+
+	m := map[*snap.Info]*wrappers.AddSnapServicesOptions{
+		info: nil,
+	}
+
+	modified, err := wrappers.EnsureSnapServices(m, progress.Null)
+	c.Assert(err, IsNil)
+	c.Check(s.sysdLog, DeepEquals, [][]string{
+		{"daemon-reload"},
+	})
+	c.Assert(modified, HasLen, 1)
+	for modifiedSn, modifiedApps := range modified {
+		// don't try a DeepEquals on this, it could be a recursive data
+		// structure which go-check doesn't handle very well
+		c.Assert(modifiedSn.SnapName(), Equals, info.SnapName())
+		c.Assert(modifiedApps, HasLen, 1)
+		c.Assert(modifiedApps, DeepEquals, []*snap.AppInfo{info.Apps["svc1"]})
+	}
+
+	content, err := ioutil.ReadFile(svcFile)
+	c.Assert(err, IsNil)
+
+	dir := filepath.Join(dirs.SnapMountDir, "hello-snap", "12.mount")
+	c.Assert(string(content), Equals, fmt.Sprintf(`[Unit]
+# Auto-generated, DO NOT EDIT
+Description=Service for snap application hello-snap.svc1
+Requires=%[1]s
+Wants=network.target
+After=%[1]s network.target snapd.apparmor.service
+X-Snappy=yes
+
+[Service]
+EnvironmentFile=-/etc/environment
+ExecStart=/usr/bin/snap run hello-snap.svc1
+SyslogIdentifier=hello-snap.svc1
+Restart=on-failure
+WorkingDirectory=%[2]s/var/snap/hello-snap/12
+ExecStop=/usr/bin/snap run --command=stop hello-snap.svc1
+ExecStopPost=/usr/bin/snap run --command=post-stop hello-snap.svc1
+TimeoutStopSec=30
+Type=forking
+
+[Install]
+WantedBy=multi-user.target
+`,
+		systemd.EscapeUnitNamePath(dir),
+		dirs.GlobalRootDir,
+	))
+}
+
+func (s *servicesTestSuite) TestEnsureSnapServicesChanges(c *C) {
```

Comment:
Should we also have a test where the info changes so that services change or get removed. Or is this out of scope for this helper? Something like the test below (which fails currently): 
```
func (s *servicesTestSuite) TestEnsureSnapServicesAddsAndRemoves(c *C) {
	// create a service first
	info := snaptest.MockSnap(c, packageHello, &snap.SideInfo{Revision: snap.R(1)})
	svcFile := filepath.Join(s.tempdir, "/etc/systemd/system/snap.hello-snap.svc1.service")
	m := map[*snap.Info]*wrappers.AddSnapServicesOptions{
		info: nil,
	}
	modified, err := wrappers.EnsureSnapServices(m, progress.Null)
	c.Assert(err, IsNil)
	c.Assert(modified, HasLen, 1)
	c.Assert(svcFile, testutil.FilePresent)

	// now simulate a refresh to a package with a different services
	// (no svc1 anymore but svc2)
	svcFileForSvc2 := filepath.Join(s.tempdir, "/etc/systemd/system/snap.hello-snap.svc2.service")
	const packageHelloSvc2 = `name: hello-snap
version: 2.71
summary: hello
description: Hello...
apps:
 svc2:
  command: bin/svc2
  daemon: simple
`
	infoSvc2 := snaptest.MockSnap(c, packageHelloSvc2, &snap.SideInfo{Revision: snap.R(32)})
	mSvc2 := map[*snap.Info]*wrappers.AddSnapServicesOptions{
		infoSvc2: nil,
	}
	modifiedSvc2, err := wrappers.EnsureSnapServices(mSvc2, progress.Null)
	c.Assert(err, IsNil)
	c.Check(s.sysdLog, DeepEquals, [][]string{
		{"daemon-reload"},
	})
	c.Check(modifiedSvc2, HasLen, 1)
	c.Check(svcFile, testutil.FileAbsent)
	c.Check(svcFileForSvc2, testutil.FilePresent)

	// now simulate a refresh to a package with no services
	const packageHelloEmpty = `name: hello-snap
version: 3.14
summary: hello
description: Hello...
`
	infoEmpty := snaptest.MockSnap(c, packageHelloEmpty, &snap.SideInfo{Revision: snap.R(64)})
	mEmpty := map[*snap.Info]*wrappers.AddSnapServicesOptions{
		infoEmpty: nil,
	}
	modifiedEmpty, err := wrappers.EnsureSnapServices(mEmpty, progress.Null)
	c.Assert(err, IsNil)
	c.Check(s.sysdLog, DeepEquals, [][]string{
		{"daemon-reload"},
	})
	c.Check(modifiedEmpty, HasLen, 1)
	c.Check(svcFile, testutil.FileAbsent)
	c.Check(svcFileForSvc2, testutil.FileAbsent)
}
```

Reply:
As mentioned, this function doesn't really need to handle service removal, and it would be a fair amount of complexity to handle it, so I will leave it out. But we can add a test about a snap that has an existing service that doesn't get changed and running EnsureSnapServices when there is a new service added to indicate that it properly handles new service units being written.

Reply:
I added some similar tests (not about removing snap services, but for snaps with multiple units where only one of them was modified, etc.)